SpannerIO: support max commit delay

kberezin-nshl opened the pull request:
Addresses https://github.com/apache/beam/issues/31007

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/workflows/Build%20python%20source%20distribution%20and%20wheels/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/workflows/Python%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/workflows/Java%20Tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[![Go tests](https://github.com/apache/beam/workflows/Go%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

kberezin-nshl commented:
R: @damccorm

kberezin-nshl commented:
assign set of reviewers

kberezin-nshl commented:
@nielm hey, sorry for tagging you explicitly here but I think I have accidentally added a wrong reviewer to this PR. Could you please help me assign the right people here?

nielm commented:
Effectively we already have this at the RPC level

```java
  /** Specifies the commit deadline. This is overridden if the CommitRetrySettings is specified. */
  public SpannerConfig withCommitDeadline(Duration commitDeadline) {
    return withCommitDeadline(ValueProvider.StaticValueProvider.of(commitDeadline));
  }
    /** Specifies the commit retry settings. Setting this overrides the commit deadline. */
  public SpannerConfig withCommitRetrySettings(RetrySettings commitRetrySettings) {
    return toBuilder().setCommitRetrySettings(commitRetrySettings).build();
  }
```

Both of these allow the deadline for a commit to complete to be set (The [default is 15s)](https://github.com/apache/beam/blame/master/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerConfig.java#L46C2-L48C35). 

Setting the commit deadline to significantly longer increases the risk of Spanner going into overload and pushing back against writes, which will reduce the throughput of the pipeline.

kberezin-nshl commented:
@nielm oh, I believe that is a completely different feature: [documentation link](https://cloud.google.com/spanner/docs/throughput-optimized-writes#set_mixed_commit_delay_times).

It was only release on March 26th, a couple of weeks ago: [release notes link](https://cloud.google.com/spanner/docs/release-notes#March_26_2024)

kberezin-nshl commented:
@nielm did you have a chance to look at this again?

nielm commented:
The setting of the commit deadline timeout predates the per-transactiom commit delay feature, but it has the same effect - limiting the duration that the spanner server can spend processing a commit. The difference is that the commit delay can be applied to different transactions, while the commit deadline is global for a spanner client. However in beam it is not possible to set the commit delay per transaction, so it is also global.

 There is no benefit to adding the commit delay to beam, when the commit deadline already exists. And the commit deadline will act as a maximum commit delay.

kberezin-nshl commented:
@nielm I am sorry but this is a _completely different_ feature. Please read the documentation link I am referring to. It was released a couple of weeks ago and we are already using it in production (we had to fork `SpannerIO` to be able to do so) and it improved performance of our jobs which write millions of records to Spanner dramatically.

Setting commit deadline does not affect performance of the writes at all, it just tells the client library how much time it can wait and retry the transaction, whereas setting a maximum commit delay may improve performance significantly if your app doesn't care much about the exact commit timing. Documentation quote:
> If you have a latency tolerant application and want to optimize throughput, setting a longer commit delay time significantly improves throughput while incurring higher latency for each write.

kberezin-nshl commented:
@johnjcasey may I ask you to help? I think we got a little confused here.
We really need this feature merged in, as we already use it and it proven to be extremely effective.

nielm commented:
My apologies - I am indeed confusing it with the commit timeout delay...

Two comments on the PR 

- can you add tests in the Write test code not the Read tests to verify that the parameter is correctly passed to writeAtLeastOnce().
- do you have any suggestion for a default value?

kberezin-nshl commented:
> can you add tests in the Write test code not the Read tests to verify that the parameter is correctly passed to writeAtLeastOnce()

Ah, sure. No idea why I updated `Read` instead of `Write` 🤦  
Done.

> do you have any suggestion for a default value?

I think by default we should just leave it out (as I implemented), because there is already a [documented default behavior](https://cloud.google.com/spanner/docs/throughput-optimized-writes#default-behavior) for this case:
> If you don't set a commit delay time, Spanner might set a small delay for you if it thinks that will amortize the cost of your writes.

kberezin-nshl commented:
There is a test failure, but I think it is unrelated to this PR:
```
org.apache.beam.sdk.io.gcp.datastore.RampupThrottlingFnTest > testRampupThrottler FAILED
    java.lang.AssertionError at RampupThrottlingFnTest.java:107
        Caused by: org.mockito.exceptions.verification.TooManyActualInvocations at RampupThrottlingFnTest.java:53
```

nielm reviewed (approved):
LGTM

nielm commented:
assign set of reviewers

nielm commented:
R: @chamikaramj for IO

kberezin-nshl commented:
@nielm @chamikaramj Hi, can we get this in, please? It's been 1.5 months.

nielm commented:
Run Java PreCommit

liferoad commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerConfig.java:
```diff
@@ -74,6 +74,8 @@ public abstract class SpannerConfig implements Serializable {
 
   public abstract @Nullable ValueProvider<RpcPriority> getRpcPriority();
 
+  public abstract @Nullable ValueProvider<Duration> getMaxCommitDelay();
```
is this worth calling out in CHANGES.md?

chamikaramj reviewed (commented):
Thanks. Just one clarifying question in addition to what XQ mentioned.

nielm commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java:
```diff
@@ -2256,6 +2264,21 @@ private void spannerWriteWithRetryIfSchemaChange(List<Mutation> batch) throws Sp
+      return Stream.of(
+              spannerConfig.getRpcPriority() != null && spannerConfig.getRpcPriority().get() != null
+                  ? Options.priority(spannerConfig.getRpcPriority().get())
+                  : null,
```
I don't see what you  mean here. This code is refactoring setting the transaction options to a separate function, and adding the `MaxCommitDelay` oprion. 

In the old code, (L2230) if there is a `RpcPriority` value set, it calls 
   `.writeAtLeastOnceWithOptions(batch, Options.priority(spannerConfig.getRpcPriority().get()));`
and if not, 
   `spannerAccessor.getDatabaseClient().writeAtLeastOnce(batch);`

In the new code it always calls 
   `.writeAtLeastOnceWithOptions(batch,getTransactionOptions())`

 where `getTransactionOptions())` returns an array of non-null `Option.TransactionOption` containing `RpcPriority` and/or `MaxCommitDelay` if they are set.

chamikaramj replied:
Thanks. For clarifying.

chamikaramj reviewed (commented):
LGTM.

chamikaramj commented:
In the failed test suite Spanner tests passed: https://ge.apache.org/s/n4qn2ounkz6b2/tests/overview

So seems like it's unrelated.
